workshops/models.py

```diff
@@ -263,6 +263,33 @@ def get_by_natural_key(self, username):
         else:
             return super().get_by_natural_key(username)
 
+    def annotate_with_instructor_eligibility(self):
+        def passed(requirement):
+            return Sum(Case(When(trainingprogress__requirement__name=requirement,
+                                 trainingprogress__state='p',
+                                 trainingprogress__discarded=False,
+                                 then=1),
+                            default=0,
+                            output_field=IntegerField()))
+
+        return self.annotate(
+            passed_training=passed('Training'),
+            passed_swc_homework=passed('SWC Homework'),
+            passed_dc_homework=passed('DC Homework'),
+            passed_discussion=passed('Discussion'),
+            passed_swc_demo=passed('SWC Demo'),
+            passed_dc_demo=passed('DC Demo'),
+        ).annotate(
+            swc_eligible=(F('passed_training') *
+                          F('passed_swc_homework') *
+                          F('passed_discussion') *
+                          F('passed_swc_demo')),
+            dc_eligible=(F('passed_training') *
+                         F('passed_dc_homework') *
+                         F('passed_discussion') *
+                         F('passed_dc_demo')),
```

Comment:
Yes, but I don't know how we can improve it (without using `reduce`, which makes code much harder to understand).